In [2]:
from IPython.display import display

import math
import sympy
import scipy
import numpy as np

# x, y, z = symbols('x y z')
sympy.init_printing(use_unicode=True)

# Homework

## Exercise set 6.1

Use the Gaussian Elimination Algorithm to solve the following linear systems, if possible, and determine whether row interchanges are necessary.

### 6a

$\begin{aligned}
x_2 - 2x_3 = 4 \\
x_1 - x_2 + x_3 = 6 \\
x_1 - x_3 = 2 \\
\end{aligned}$

**solution**

$\begin{bmatrix}
0 & 1 & -2 & 4 \\
1 & -1 & 1 & 6 \\
1 & 0 & -1 & 2 \\
\end{bmatrix}$

$E_1 + E2 \rightarrow E_1$

$\begin{bmatrix}
1 & 0 & -1 & 10 \\
1 & -1 & 1 & 6 \\
1 & 0 & -1 & 2 \\
\end{bmatrix}$

$-E_2 + E3 \rightarrow E_2$

$\begin{bmatrix}
1 & 0 & -1 & 10 \\
0 & 1 & -2 & -4 \\
1 & 0 & -1 & 2 \\
\end{bmatrix}$

$E_3 - E1 \rightarrow E_3$

$\begin{bmatrix}
1 & 0 & -1 & 10 \\
0 & 1 & -2 & -4 \\
0 & 0 & 0 & -8 \\
\end{bmatrix}$




In [3]:
m = sympy.Matrix([[0, 1, -2, 4], [1, -1, 1, 6], [1, 0, -1, 2]])
m.rref()

⎛⎡1  0  -1  0⎤           ⎞
⎜⎢           ⎥           ⎟
⎜⎢0  1  -2  0⎥, (0, 1, 3)⎟
⎜⎢           ⎥           ⎟
⎝⎣0  0  0   1⎦           ⎠

### 6c

$\begin{aligned}
2x_1 - x_2 + x_3 - x_4 = 6 \\
x_2 - x_3 + x_4 = 5 \\
x_4 = 5 \\
x_3 - x_4 = 3 \\
\end{aligned}$

**solution**

$\begin{bmatrix}
2 & -1 & 1 & -1 & 6 \\
0 & 1 & -1 & 1 & 5 \\
0 & 0 & 0 & 1 & 5 \\
0 & 0 & 1 & -1 & 3 \\
\end{bmatrix}$

$E_1 + E_2 \rightarrow E_1$

$\begin{bmatrix}
2 & 0 & 0 & 0 & 11 \\
0 & 1 & -1 & 1 & 5 \\
0 & 0 & 0 & 1 & 5 \\
0 & 0 & 1 & -1 & 3 \\
\end{bmatrix}$

$E_2 + E_4 \rightarrow E_2$

$\begin{bmatrix}
2 & 0 & 0 & 0 & 11 \\
0 & 1 & 0 & 0 &  8 \\
0 & 0 & 0 & 1 & 5 \\
0 & 0 & 1 & -1 & 3 \\
\end{bmatrix}$

$E_3 + E_4 \rightarrow E_3$

$\begin{bmatrix}
2 & 0 & 0 & 0 & 11 \\
0 & 1 & 0 & 0 & 8 \\
0 & 0 & 1 & 0 & 8 \\
0 & 0 & 1 & -1 & 3 \\
\end{bmatrix}$

$E_3 - E_4 \rightarrow E_4$

$\begin{bmatrix}
2 & 0 & 0 & 0 & 11 \\
0 & 1 & 0 & 0 & 8 \\
0 & 0 & 1 & 0 & 8 \\
0 & 0 & 0 & 1 & 5 \\
\end{bmatrix}$

$x_4 = 5, x_3=13, x_2=13, x_1=\frac{11}{2}$

In [39]:
m = sympy.Matrix([
    [2, -1, 1, -1, 6],
    [0, 1, -1, 1, 5],
    [0, 0, 0, 1, 5],
    [0, 0, 1, -2, 3],
])
display(m.rref())
np.array(m)[:,:-1] @ np.array([11/2, 13, 13, 5])

⎛⎡1  0  0  0  11/2⎤              ⎞
⎜⎢                ⎥              ⎟
⎜⎢0  1  0  0   13 ⎥              ⎟
⎜⎢                ⎥, (0, 1, 2, 3)⎟
⎜⎢0  0  1  0   13 ⎥              ⎟
⎜⎢                ⎥              ⎟
⎝⎣0  0  0  1   5  ⎦              ⎠

array([6.00000000000000, 5.00000000000000, 5.00000000000000,
       3.00000000000000], dtype=object)

### 14d

Repeat exercise 7 using the Gauss-Jordan method.

Use Algorithm 6.1 and Maple with $Digits:=10$ to solve the following linear systems.

$\begin{aligned}
2x_1 + x_2 - x_3 + x_4 - 3x_5 = 7 \\
x_1 + 2x_3 - x_4 + x_5 = 2 \\
-2x_2 - x_3 + x_4 - x_5 = -5 \\
3x_1 + x_2 - 4x_3 + 5x_5 = 6 \\
x_1 - x_2 - x_3 - x_4 + x_5 = 3 \\
\end{aligned}$

**solution**

$x_1=\frac{328}{171}, x_2=\frac{112}{57}, x_3=-\frac{169}{171}, x_4=-\frac{182}{57}, x_5=-\frac{194}{171}$

In [41]:
m = sympy.Matrix([
    [2, 1, -1, 1, -3, 7],
    [1, 0, 2, -1, 1, 2],
    [0, -2, -1, 1, -1, -5],
    [3, 1, -4, 0, 5, 6],
    [1, -1, -1, -1, 1, 3],
])
sol = m.rref()[0].col(5)
display(sol)
# Sanity check
np.array(m)[:,:-1] @ np.array(sol)

⎡ 328 ⎤
⎢ ─── ⎥
⎢ 171 ⎥
⎢     ⎥
⎢ 112 ⎥
⎢ ─── ⎥
⎢  57 ⎥
⎢     ⎥
⎢-169 ⎥
⎢─────⎥
⎢ 171 ⎥
⎢     ⎥
⎢-182 ⎥
⎢─────⎥
⎢  57 ⎥
⎢     ⎥
⎢-194 ⎥
⎢─────⎥
⎣ 171 ⎦

array([[7],
       [2],
       [-5],
       [6],
       [3]], dtype=object)

## Exercise set 6.2

### 10b

Use Gaussian elimination and three-digit chopping arithmetic to solve the following linear systems, and compare the approximations to the actual solution.

$\begin{aligned}
3.3330x_1 + 15920x_2 + 10.333x_3 = 7953 \\
2.2220x_1 + 16.710x_2 + 9.6120x_3 = 0.965 \\
-1.5611x_1 + 5.1791x_2 - 1.6855x_3 = 2.714 \\
\end{aligned}$

Actual solution $[1, 0.5, -1]$.

**solution**



In [54]:
data = [
    [3.3330, 15920, 10.333, 7953],
    [2.2220, 16.710, 9.6120, 0.965],
    [-1.5611, 5.1791, -1.6852, 2.714],
]

m = np.array(data)
sol = np.linalg.solve(m[:,:-1], m[:,-1])
print(sol)
print(m[:,:-1] @ np.array([1, 0.5, -1]))

[ 0.99970137  0.50000002 -0.999931  ]
[7.95300e+03 9.65000e-01 2.71365e+00]


### 14b

Repeat Exercise 10 using Gaussian elimination with partial pivoting.

**solution**



### 18b

Repeat Exercise 10 using Gaussian elimination with scaled partial pivoting.

**solution**

$x_1=0.9997, x_2 = 0.5000, x_3 = -0.9999$

In [55]:
data = [
    [3.3330, 15920, 10.333, 7953],
    [2.2220, 16.710, 9.6120, 0.965],
    [-1.5611, 5.1791, -1.6852, 2.714],
]

m = np.array(data)
sol = np.linalg.solve(m[:,:-1], m[:,-1])
print(sol)
print(m[:,:-1] @ np.array([1, 0.5, -1]))

[ 0.99970137  0.50000002 -0.999931  ]
[7.95300e+03 9.65000e-01 2.71365e+00]


## Exercise set 6.5

### 2a

Solve the following linear systems:

$\begin{bmatrix}
1 & 0 & 0 \\
-2 & 1 & 0 \\
3 & 0 & 1 \\
\end{bmatrix}
\quad
\begin{bmatrix}
2 & 1 & -1 \\
0 & 4 & 2 \\
0 & 0 & 5 \\
\end{bmatrix}
\quad
\begin{bmatrix}
x_1 \\
x_2 \\
x_3 \\
\end{bmatrix}
\quad
=
\quad
\begin{bmatrix}
1 \\
0 \\
5 \\
\end{bmatrix}$

**solution**

$Ly=
\quad
\begin{bmatrix}
1 & 0 & 0 \\
-2 & 1 & 0 \\
3 & 0 & 1 \\
\end{bmatrix}
\quad
\begin{bmatrix}
y_1 \\
y_2 \\
y_3 \\
\end{bmatrix}
\quad
=
\quad
\begin{bmatrix}
1 \\
0 \\
5 \\
\end{bmatrix}$

$y=
\quad
\begin{bmatrix}
1 \\
2 \\
-3 \\
\end{bmatrix}$

$\boldsymbol{b}=L\boldsymbol{y}=
\quad
\begin{bmatrix}
1 & 0 & 0 \\
-2 & 1 & 0 \\
3 & 0 & 1 \\
\end{bmatrix}
\quad
\begin{bmatrix}
1 \\
2 \\
-3 \\
\end{bmatrix}
\quad
=
\quad
\begin{bmatrix}
1 \\
0 \\
0 \\
\end{bmatrix}$

$U\boldsymbol{x}=\boldsymbol{y}=
\quad
\begin{bmatrix}
2 & 1 & -1 \\
0 & 4 & 2 \\
0 & 0 & 5 \\
\end{bmatrix}
\quad
\begin{bmatrix}
x_1 \\
x_2 \\
x_3 \\
\end{bmatrix}
\quad
=
\quad
\begin{bmatrix}
1 \\
0 \\
0 \\
\end{bmatrix}$

$x_1 = \frac{3}{4}, x_2 = \frac{3}{10}, x_3 = \frac{2}{5}$


In [79]:
L = np.array([
    [1, 0, 0],
    [-2, 1, 0],
    [3, 0, 1],
])
y = np.array([1, 2, -3])
Ly = L @ y
print(Ly)
U = np.array([
    [2, 1, -2],
    [0, 4, 2],
    [0, 0, 5],
])
np.linalg.solve(L@U, np.array([1, 0, 5]))
L@U@np.array([3/4, 0.3, 0.4])

[1 0 0]


array([1.00000000e+00, 2.22044605e-16, 5.00000000e+00])

### 6a

Factor the following matrices into the LU decomposition using the LU Factorization Algorithm with $l_{ii} = 1$ for all $i$.

$\begin{bmatrix}
1 & -1 & 0 \\
2 & 2 & 3 \\
-1 & 3 & 2 \\
\end{bmatrix}$

**solution**

$LU=
\quad
\begin{bmatrix}
1 & 0 & 0 \\
-\frac{1}{2} & 1  & 0 \\
\frac{1}{2} & -\frac{1}{2} & 1 \\
\end{bmatrix}
\quad
\begin{bmatrix}
2 & 2 & 3 \\
0 & 4 & \frac{7}{2} \\
0 & 0 & \frac{1}{4} \\
\end{bmatrix}$

In [71]:
m = np.array([
    [1, -1, 0],
    [2, 2, 3],
    [-1, 3, 2],
])
lu, piv = scipy.linalg.lu_factor(m)
L, U = np.tril(lu, k=-1) + np.eye(3), np.triu(lu)
print(L)
print(U)

[[ 1.   0.   0. ]
 [-0.5  1.   0. ]
 [ 0.5 -0.5  1. ]]
[[2.   2.   3.  ]
 [0.   4.   3.5 ]
 [0.   0.   0.25]]


## Exercise set 6.6

### 2

Determine which of the following matrices are (i) symmetric, (ii) singular, (iii) stricly diagonally dominant, (iv) positive definite.

**solution**

a = (iv)

b = (iii)

c = (i)

d = (ii)

### 4a

Use the $LDL^t$ Factorization Algorithm to find a factorization of the form $A = LDL^t$ for the following matrices:

$\begin{bmatrix}
4 & -1 & 1 \\
-1 & 3 & 0 \\
1 & 0 & 2 \\
\end{bmatrix}$

**solution**

$L=
\quad
\begin{bmatrix}
1 & 0 & 0 \\
-0.25 & 1 & 0 \\
0.25 & 0.0909 & 1 \\
\end{bmatrix}$

$D=
\quad
\begin{bmatrix}
4 & 0 & 0 \\
0 & 2.75 & 0 \\
0 & 0 & 1.72727 \\
\end{bmatrix}$

In [90]:
m = np.array([
    [4, -1, 1],
    [-1, 3, 0],
    [1, 0, 2],
])
scipy.linalg.ldl(m, lower=True)

(array([[ 1.        ,  0.        ,  0.        ],
        [-0.25      ,  1.        ,  0.        ],
        [ 0.25      ,  0.09090909,  1.        ]]),
 array([[4.        , 0.        , 0.        ],
        [0.        , 2.75      , 0.        ],
        [0.        , 0.        , 1.72727273]]),
 array([0, 1, 2]))

### 6a

Use the Cholesky Algorithm to find a factorization of the form $A=LL^t$ for the matrices in Exercise 4.

**solution**

$L=
\quad
\begin{bmatrix}
2 & 0 & 0 \\
-\frac{1}{2} & 1.65813 & 0 \\
\frac{1}{2} & 0.15076 & 1.31426 \\
\end{bmatrix}$

In [89]:
m = np.array([
    [4, -1, 1],
    [-1, 3, 0],
    [1, 0, 2],
])
scipy.linalg.cholesky(m, lower=True)

array([[ 2.        ,  0.        ,  0.        ],
       [-0.5       ,  1.6583124 ,  0.        ],
       [ 0.5       ,  0.15075567,  1.31425748]])

### 12c

Use Crout factorization for tridiagnonal systems to solve the following linear systems.

$\begin{aligned}
2x_1 - x_2 = 3 \\
x_1 + 2x_2 - x_3 = 4 \\
x_2 - 2x_3 + x_4 = 0 \\
x_3 + 2x_4 = 6 \\
\end{aligned}$

**solution**

$LU=
\quad
\begin{bmatrix}
1 & 0 & 0 & 0 \\
0.5 & 1 & 0 & 0 \\
0 & 0.4 & 1 & 0 \\
0 & 0 & -0.625 & 1 \\
\end{bmatrix}
\quad
\begin{bmatrix}
2 & -1 & 0 & 0 \\
0 & 2.5 & -1 & 0 \\
0 & 0 & -1.6 & 1 \\
0 & 0 & 0 & 2.625 \\
\end{bmatrix}$

$\boldsymbol{b}=L\boldsymbol{y}$

$\boldsymbol{y}=
\begin{bmatrix}
3 \\
2.5 \\
-1 \\
5.375 \\
\end{bmatrix}$

$U\boldsymbol{x}=\boldsymbol{y}$

$x_1=2.38095, x_2 = 1.76190, x_3 = 1.90476, x_4 = 2.047619$

In [98]:
m = np.array([
    [2, -1, 0, 0],
    [1, 2, -1, 0],
    [0, 1, -2, 1],
    [0, 0, 1, 2],
])
lu, piv = scipy.linalg.lu_factor(m)
L, U = np.tril(lu, k=-1) + np.eye(4), np.triu(lu)
print(L)
print(U)

y = np.linalg.solve(L, np.array([3, 4, 0, 6]))
print(y)
x = np.linalg.solve(U, y)
print(x)
m @ x

[[ 1.     0.     0.     0.   ]
 [ 0.5    1.     0.     0.   ]
 [ 0.     0.4    1.     0.   ]
 [ 0.     0.    -0.625  1.   ]]
[[ 2.    -1.     0.     0.   ]
 [ 0.     2.5   -1.     0.   ]
 [ 0.     0.    -1.6    1.   ]
 [ 0.     0.     0.     2.625]]
[ 3.     2.5   -1.     5.375]
[2.38095238 1.76190476 1.9047619  2.04761905]


array([3., 4., 0., 6.])